### PR TITLE
DRILL-4185: UNION ALL involving empty directory on any side of union …

### PR DESCRIPTION
**Pull request by vdiravka**
…all results in Failed query
SchemaLessTable, SchemaLessScan, SchemaLessBatchCreator, SchemaLessBatch classes are inroduced. 
The main idea that empty directory is a Drill (SchemaLess) table.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/SchemalessScan.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @Override
+  public int getMaxParallelizationWidth() {
+    return 1;
+  }
+
+  @Override
+  public String getDigest() {
+    return toString();
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/SchemalessScan.java]: What value `toString` will return?

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/RefreshMetadataHandler.java [hunk 2]
```diff
@@ -78,19 +78,24 @@ public PhysicalPlan getPlan(SqlNode sqlNode) throws ValidationException, RelConv
 
       final Table table = schema.getTable(tableName);
 
-      if(table == null){
+      if(table == null) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/RefreshMetadataHandler.java]: Please add space before `if`, here and below.

File: exec/java-exec/src/test/java/org/apache/drill/test/BaseTestQuery.java [hunk 3]
```diff
@@ -119,6 +125,15 @@ public static void setupDefaultTestCluster() throws Exception {
     // turns on the verbose errors in tests
     // sever side stacktraces are added to the message before sending back to the client
     test("ALTER SESSION SET `exec.errors.verbose` = true");
+    emptyDirCreating();
+  }
+
+  /**
+   * Creates an empty directory under <b>dfs.root</b> schema.
+   */
+  private static void emptyDirCreating() {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/test/BaseTestQuery.java]: Please rename -> `createEmptyDirectory`.

File: exec/java-exec/src/test/java/org/apache/drill/test/BaseTestQuery.java [hunk 4]
```diff
@@ -119,6 +125,15 @@ public static void setupDefaultTestCluster() throws Exception {
     // turns on the verbose errors in tests
     // sever side stacktraces are added to the message before sending back to the client
     test("ALTER SESSION SET `exec.errors.verbose` = true");
+    emptyDirCreating();
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/test/BaseTestQuery.java]: Why do we create empty directory for all tests that extend this class? I guess need to create it only for those tests that need it.

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestParquetGroupScan.java [hunk 5]
```diff
@@ -56,65 +56,50 @@ private void prepareTables(final String tableName, boolean refreshMetadata) thro
...
     prepareTables("4376_1", true);
 
-    testBuilder()
-      .sqlQuery("SELECT COUNT(*) AS `count` FROM dfs.tmp.`4376_1/60*`")
-      .ordered()
-      .baselineColumns("count").baselineValues(1984L)
-      .go();
+    int actualRecordCount = testSql("SELECT * FROM dfs.tmp.`4376_1/60*`");
+    int expectedRecordCount = 1984;
+    assertEquals(String.format("Received unexpected number of rows in output: expected=%d, received=%s",
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestParquetGroupScan.java]: `expected=%d` -> `expected = %d`. Please correct here and in other cases.

File: exec/java-exec/src/test/java/org/apache/drill/exec/TestEmptyInputSql.java [hunk 6]
```diff
@@ -177,4 +177,33 @@ public void testQueryEmptyCsv() throws Exception {
...
+        .schemaBaseLine(expectedSchema)
+        .build()
+        .run();
+  }
+
+  @Test
+  public void testEmptyDirectoryAndFieldInQuery() throws Exception {
+    final List<Pair<SchemaPath, TypeProtos.MajorType>> expectedSchema = Lists.newArrayList();
+    final TypeProtos.MajorType majorType = TypeProtos.MajorType.newBuilder()
+        .setMinorType(TypeProtos.MinorType.INT) // field "key" is absent in schema-less table
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/TestEmptyInputSql.java]: schema-less -> schemaless

File: exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java [hunk 7]
```diff
@@ -1197,4 +1197,64 @@ public void testFieldWithDots() throws Exception {
       .baselineValues("1", "2", "1", null, "a")
       .go();
   }
-}
\ No newline at end of file
+
+  @Test
+  public void testUnionAllRightEmptyDir() throws Exception {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/TestUnionAll.java]: 1. Does union (aka union distinct) also work fine?
2. Case when bot directories in union / union all are empty?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java [hunk 8]
```diff
@@ -250,20 +250,12 @@ private boolean metaDataFileExists(FileSystem fs, FileStatus dir) throws IOExcep
...
-
-          if (metaDataFileExists(fs, dir)) {
-            return true;
-          }
-          List<FileStatus> statuses = DrillFileSystemUtil.listFiles(fs, dir.getPath(), false);
-          return !statuses.isEmpty() && super.isFileReadable(fs, statuses.get(0));
-        }
+        // There should be at least one file, which is readable by Drill
+        List<FileStatus> statuses = DrillFileSystemUtil.listFiles(fs, dir.getPath(), false);
+        return !statuses.isEmpty() && super.isFileReadable(fs, statuses.get(0));
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: Why did you remove `metaDataFileExists(fs, dir)` check? `DrillFileSystemUtil` won't list dot files.
 Plus it seems you break the assumptions highlighted in `isReadable` method above.

Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: I did it on purpose. With the old logic of isDirReadable() method an empty directory, which contains parquet metadata files, will be processes with ParquetGroupScan as a Parquet Table. It leads to obtaining an exception:
https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetGroupScan.java#L878

To process such table with SchemalessScan, isReadable method should return false for that case. In other words it shouldn't check availability of metadata cache files, but only really readable files by Drill.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/SchemalessScan.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+import java.util.List;
+
+/**
+ *  The type of scan operator, which allows to scan schemaless tables ({@link DynamicDrillTable} with null selection)
+ */
+@JsonTypeName("schemaless-scan")
+public class SchemalessScan extends AbstractGroupScan implements SubScan {
+
+  private String selectionRoot;
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/base/SchemalessScan.java]: final

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java [hunk 10]
```diff
@@ -424,6 +429,23 @@ public MetadataContext getMetaContext() {
...
+  public boolean isEmptyDirectory() {
+    return emptyDirectory;
+  }
+
+  /**
+   * Setting this as true allows to identify this as empty directory file selection
+   *
+   * @param emptyDirectory empty directory flag
+   */
+  public void setEmptyDirectory(boolean emptyDirectory) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java]: Please use set empty directory without flag.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java (hunk 8 above)
Comment (arina-ielchiieva) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: 1. How we know that we have file in nested directory if filter checks with recursive flag false?
2. The `isReadable` method uses implicit assumption that if metadata file exists, there is some data in folder. Plus comment in this file also  points out that the same logic is used in `isDirReadable`. Since you are removing this logic from `isDirReadable` please make sure the logic is synced between two methods or at least comment is updated to reflect new approach.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java (hunk 10 above)
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSelection.java]: Done

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java (hunk 8 above)
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/ParquetFormatPlugin.java]: I have reverted my changes here. SchemalessScan will be created if ParquetGroupScan has empty entries and can't be used. The info that parquet metadata files are invalid will be logged.

File: exec/java-exec/src/test/java/org/apache/drill/TestJoinNullable.java [hunk 11]
```diff
@@ -568,6 +570,22 @@ public void nullMixedComparatorEqualJoinHelper(final String query) throws Except
         .go();
   }
 
+  /** InnerJoin with empty dir table on nullable cols, MergeJoin */
+  // TODO: the same tests should be added for HashJoin operator, DRILL-6070
+  @Test
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/TestJoinNullable.java]: Please add test for merge join with ignore annotation. Also please check that NLJ also works fine. If not, please fix.

Comment (vdiravka) [reply exec/java-exec/src/test/java/org/apache/drill/TestJoinNullable.java]: The bug was founded for NLJ and empty tables. I have resolved that issue.
The separate test class is added for empty dir tables and different join operators.

Also I have made refactoring for the TestHashJoinAdvanced, TestMergeJoinAdvanced, TestNestedLoopJoin classes.
